Disable Vim's default/built-in digraphs.

diff --git a/src/digraph.c b/src/digraph.c
index dd5920a24..7bef9f9d2 100644
--- src/digraph.c
+++ src/digraph.c
@@ -1865,41 +1865,11 @@ get_digraph_for_char(int val_arg)
 {
     int		val = val_arg;
     int		i;
-    int		use_defaults;
     digr_T	*dp;
     static      char_u      r[3];
 
-#if defined(USE_UNICODE_DIGRAPHS)
-    if (!enc_utf8)
-    {
-	char_u	    buf[6], *to;
-	vimconv_T   vc;
-
-	// convert the character from 'encoding' to Unicode
-	i = mb_char2bytes(val, buf);
-	vc.vc_type = CONV_NONE;
-	if (convert_setup(&vc, p_enc, (char_u *)"utf-8") == OK)
-	{
-	    vc.vc_fail = TRUE;
-	    to = string_convert(&vc, buf, &i);
-	    if (to != NULL)
-	    {
-		val = utf_ptr2char(to);
-		vim_free(to);
-	    }
-	    (void)convert_setup(&vc, NULL, NULL);
-	}
-    }
-#endif
-
-    for (use_defaults = 0; use_defaults <= 1; use_defaults++)
-    {
-	if (use_defaults == 0)
 	    dp = (digr_T *)user_digraphs.ga_data;
-	else
-	    dp = digraphdefault;
-	for (i = 0; use_defaults ? dp->char1 != NUL
-					       : i < user_digraphs.ga_len; ++i)
+	for (i = 0; i < user_digraphs.ga_len; ++i)
 	{
 	    if (dp->result == val)
 	    {
@@ -1910,7 +1880,6 @@ get_digraph_for_char(int val_arg)
 	    }
 	    ++dp;
 	}
-    }
     return NULL;
 }
 
@@ -1987,22 +1956,6 @@ getexactdigraph(int char1, int char2, int meta_char)
 	++dp;
     }
 
-    /*
-     * Search default digraphs.
-     */
-    if (retval == 0)
-    {
-	dp = digraphdefault;
-	for (i = 0; dp->char1 != 0; ++i)
-	{
-	    if ((int)dp->char1 == char1 && (int)dp->char2 == char2)
-	    {
-		retval = dp->result;
-		break;
-	    }
-	    ++dp;
-	}
-    }
 #ifdef USE_UNICODE_DIGRAPHS
     if (retval != 0 && !enc_utf8)
     {
